fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -134,6 +163,98 @@ impl<T> Interpreter<PredicateStorage, T> {
 
+    async fn verify_predicate_async<Tx, E>(
+        tx: Tx,
+        balances: InitialBalances,
+        params: ConsensusParameters,
+        gas_costs: GasCosts,
+    ) -> Result<PredicatesChecked, PredicateVerificationFailed>
+    where
+        Tx: ExecutableTransaction + Send + 'static,
+        E: ParallelExecutor
+            + ParallelExecutor<TaskResult = Result<Word, PredicateVerificationFailed>>,
+    {
+        if !tx.check_predicate_owners(&params.chain_id) {
+            return Err(PredicateVerificationFailed::InvalidOwner)
+        }
+
+        let mut verifications = vec![];
+
+        for index in 0..tx.inputs().len() {
+            let is_predicate = matches!(
+                tx.inputs()[index],
+                Input::CoinPredicate(_)
+                    | Input::MessageCoinPredicate(_)
+                    | Input::MessageDataPredicate(_)
+            );
+
+            if !is_predicate {
+                continue
+            }
+
+            if let Some(predicate) = RuntimePredicate::from_tx(&params, &tx, index) {
+                let gas_costs = gas_costs.clone();
+                let balances = balances.clone();
+
+                let tx = tx.clone();
+
+                let verify_task = E::create_task(move || {
+                    let mut vm = Interpreter::with_storage(
+                        PredicateStorage::default(),
+                        params,
+                        gas_costs,
+                    );
+
+                    let context = Context::PredicateVerification { program: predicate };
+
+                    let available_gas =
+                        if let Some(x) = tx.inputs()[index].predicate_gas_used() {
+                            x
+                        } else {
+                            return Err(PredicateVerificationFailed::GasNotSpecified)
+                        };
+
+                    vm.init_predicate(context, &tx, balances.clone(), available_gas)?;
+
+                    let result = vm.verify_predicate();
+                    let is_successful = matches!(result, Ok(ProgramState::Return(0x01)));
+
+                    let gas_used = available_gas
+                        .checked_sub(vm.remaining_gas())
+                        .ok_or_else(|| Bug::new(BugId::ID004, GlobalGasUnderflow))?;
+
+                    if !is_successful {
+                        result?;
+                        return Err(PredicateVerificationFailed::False)
+                    }
+
+                    if vm.remaining_gas() != 0 {
+                        return Err(PredicateVerificationFailed::GasMismatch)
+                    }
+
+                    Ok(gas_used)
+                });
+
+                verifications.push(verify_task);
+            }
+        }
+
+        let verifications = E::execute_tasks(verifications).await;
+        let cumulative_gas_used = verifications
+            .into_iter()
+            .try_fold(0, |acc, x| Ok::<u64, PredicateVerificationFailed>(acc + x?))?;
```

Comment:
It would be best if you also handled `acc + x` overflow

Reply:
done